components/script/dom/globalscope.rs

```diff
@@ -112,15 +113,41 @@ use std::ops::Index;
 use std::rc::Rc;
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::Arc;
+use std::thread::JoinHandle;
 use time::{get_time, Timespec};
 use uuid::Uuid;
 
 #[derive(JSTraceable)]
-pub struct AutoCloseWorker(Arc<AtomicBool>);
+pub struct AutoCloseWorker {
+    /// https://html.spec.whatwg.org/multipage/#dom-workerglobalscope-closing
+    closing: Arc<AtomicBool>,
+    /// A handle to join on the worker thread.
+    join_handle: Option<JoinHandle<()>>,
+    /// A sender of control messages,
+    /// currently only used to signal shutdown.
+    control_sender: Option<Sender<()>>,
+}
 
 impl Drop for AutoCloseWorker {
+    /// <https://html.spec.whatwg.org/multipage/#terminate-a-worker>
     fn drop(&mut self) {
-        self.0.store(true, Ordering::SeqCst);
+        // Step 1.
+        self.closing.store(true, Ordering::SeqCst);
+
+        // Drop the channel to signal shutdown.
+        drop(
+            self.control_sender
+                .take()
+                .expect("No control sender to worker thread."),
+        );
+
```

Comment:
Why not send a message rather than dropping the channel?

Reply:
I tried to write this down over at https://github.com/servo/servo/issues/26502

While we use a lot of exit messages throughout the codebase, I think @matklad had a good point that relying on dropping channels is better, because it removes all sort of opportunities for leaking things accidentally, and it also enforces some sort of recursive dropping/shutting-down of resources on designs(since otherwise if it doesn't drop it won't shut-down properly). 

So for example, if the control channel is cloned somewhere, then the shut-down will fail, and that would then force us to not clone the control channel, or otherwise ensure all senders drop. 

Then this also enforces proper separation of concerns in terms of channels, since we can't take the easy route and clone let's say a control channel and pass it to some child worker, since that would break the shutdown mechanism. 

I think a good example is how currently the script-thread clones it's `control_chan: IpcSender<ConstellationControlMsg>,` and passes it to layout threads, which means we can use it as a "drop to shutdown" channel, unless we actually introduce a new channel for layout -> script communication, which actually makes sense(it's inconsistent that layout sends `ConstellationControlMsg` in some cases). 

Please let me know what you think of this.

Reply:
My personal taste would be to use shutdown messages, and reserve dropping channels for the case where the sender has panicked. This is just taste though, it'll be the same shutdown code running in either case.